### DISCUSSION TITLE
Using scratchblocks

### DISCUSSION THREAD

**Discussion by ghost** [ScratchAddons/ScratchAddons#4799] [Help]
While some may say this isn't the right place to ask this, I'm asking in relation to a new addon I'm developing. What is the best way to install the scratchblocks library into my addon, because the docs for the library don't make any sense to me. Thanks in advance for your help!

**Comment by WorldLanguages** [comment]
What addon are you working on?

**Reply by ghost** [reply]
The importing of custom blocks. I want to use the scratchblocks to illustrate what blocks the user is selecting. (Also, is there any documentation as to how to use the Modal API or modal.js)

**Reply by mxmou** [reply]
I'm not sure if scratchblocks is the best way to do that: there is probably a simpler way using the VM and a read-only Blockly workspace.

The modal API isn't documented yet. There are two ways to use it:
* Creating the modal once and showing or hiding it when it's opened or closed. You can create the modal like this:

  ```js
  const { backdrop, container, content, closeButton, open, close } = addon.tab.createModal("Example");
  backdrop.addEventListener("click", close);
  closeButton.addEventListener("click", close);
  ```
  and call `open()` when you want to show the modal.
* Creating the modal when it's opened and destroying it when it's closed. That's done like this:

  ```js
  const { backdrop, container, content, closeButton, remove } = addon.tab.createModal("Example", { isOpen: true });
  backdrop.addEventListener("click", remove);
  closeButton.addEventListener("click", remove");
  ```

`content` is the element that you can append things to to make them appear inside the modal. `container` is the modal itself which you might want to style with CSS.

**Reply by TheColaber** [reply]
There is also an API the backpack uses to create images of blocks which I haven't explored which is probably the best way to create the image. we also have an addon that does this.

**Reply by mxmou** [reply]
> There is also an API the backpack uses to create images of blocks which I haven't explored which is probably the best way to create the image. we also have an addon that does this.

That would require putting the blocks into the user's backpack first. Those images are also very small.

**Reply by ghost** [reply]
Thanks for the help! Also, with the addon I'm developing, it directly edits the project.json file (by pulling it, editing it, then replacing the user's open project with the new altered one). Is that feasible/allowed?

**Reply by ghost** [reply]
If there isn't a good alternative to using scratchblocks, can someone help me out with the setup, because I don't understand where to link everything. Thanks
